src/OpenLoco/StationManager.cpp

```diff
@@ -124,22 +130,264 @@ namespace OpenLoco::StationManager
+        {
+            townDefault,
+            townNorth,
+            townSouth,
+            townEast,
+            townWest,
+            townCentral,
+            townTransfer,
+            townHalt,
+            townValley,
+            townHeights,
+            townWoods,
+            townLakeside,
+            townExchange,
+            townAirport,
+            townOilfield,
+            townMines,
+            townDocks,
+            townAnnexe,
+            townSidings,
+            townBranch,
+            upperTown,
+            lowerTown,
+            townHeliport,
+            townForest,
+            townJunction,
+            townCross,
+            townViews,
+            townOrd1,
+            townOrd2,
+            townOrd3,
+            townOrd4,
+            townOrd5,
+            townOrd6,
+            townOrd7,
+            townOrd8,
+            townOrd9,
+            townOrd10,
+            townOrd11,
+            townOrd12,
+            townOrd13,
+            townOrd14,
+            townOrd15,
+            townOrd16,
+            townOrd17,
+            townOrd18,
+            townOrd19,
+            townOrd20,
+        };
+
+        // Bit mask for station names already used in the current town.
+        std::bitset<27> realNamesInUse{};    // ebp
+        std::bitset<20> ordinalNamesInUse{}; // edi
+        for (auto& station : stations())
+        {
+            if (!StringManager::isTownName(station.name))
+                continue;
+
+            auto nameKey = StringManager::fromTownName(station.name) - StringIds::station_town;
+            if (nameKey > StationName::townOrd20)
+                continue;
+
+            if (station.town != townId)
+                continue;
+
+            if (nameKey < StationName::townOrd1)
+                realNamesInUse.set(nameKey, true);
+            else
+                ordinalNamesInUse.set(nameKey, true);
+        }
+
+        if (mode == 1)
+        {
+            // Airport
+            if (!realNamesInUse.test(StationName::townAirport))
+                return StringManager::toTownName(StringIds::station_town_airport);
+        }
+        else if (mode == 2)
+        {
+            // Heliport
+            if (!realNamesInUse.test(StationName::townHeliport))
+                return StringManager::toTownName(StringIds::station_town_heliport);
+        }
+        else if (mode == 3)
+        {
+            // 0x0048FA00
+            auto tile = TileManager::get(Map::Pos2(position.x, position.y));
+            auto* surface = tile.surface();
+            if (surface != nullptr && surface->water() == 0)
+            {
+                // Docks
+                if (!realNamesInUse.test(StationName::townDocks))
+                    return StringManager::toTownName(StringIds::station_town_docks);
+            }
+        }
+
+        // 0x0048FA41
+        if (IndustryManager::industryNearPosition(Map::Pos2(position.x, position.y), IndustryObjectFlags::oilfield))
```

Comment:
I dont think you need to do the Map::Pos2 here as Map::Pos3 is derived from Pos2.